File: backend/app/services/workspace_service.py
from fastapi import HTTPException
from sqlalchemy.orm import Session

from app.models.user import User
from app.models.workspace import Workspace


def create_workspace(
    db: Session,
    current_user: User,
    name: str,
    description: str | None,
    color: str,
    icon: str,
):
    workspace = Workspace(
       name=name,
    description=description,
    color=color,
    icon=icon,
    user_id=current_user.id,
    )

    db.add(workspace)
    db.commit()
    db.refresh(workspace)

    return workspace


def get_user_workspaces(
    db: Session,
    current_user: User,
):
    return (
        db.query(Workspace)
        .filter(
            Workspace.user_id == current_user.id
        )
        .order_by(
            Workspace.created_at.desc()
        )
        .all()
    )


def get_workspace(
    db: Session,
    workspace_id: int,
    current_user: User,
):
    workspace = (
        db.query(Workspace)
        .filter(
            Workspace.id == workspace_id,
            Workspace.user_id == current_user.id,
        )
        .first()
    )

    if not workspace:
        raise HTTPException(
            status_code=404,
            detail="Workspace not found",
        )

    return workspace


def update_workspace(
    db: Session,
    workspace: Workspace,
    name: str | None,
    description: str | None,
    color: str | None,
    icon: str | None,
):
    if name is not None:
        workspace.name = name

    if description is not None:
        workspace.description = description

    if color is not None:
     workspace.color = color

    if icon is not None:
        workspace.icon = icon

    db.commit()
    db.refresh(workspace)

    return workspace


def delete_workspace(
    db: Session,
    workspace: Workspace,
):
    db.delete(workspace)
    db.commit()

    return {
        "message": "Workspace deleted successfully"
    }

File: frontend/src/services/workspace/workspaceService.ts
import api from "@/api/axios";

import type { Workspace } from "@/types/workspace";

export async function getWorkspaces(): Promise<Workspace[]> {
  const response = await api.get(
    "/api/workspaces",
  );

  return response.data;
}

export async function createWorkspace(
  data: {
    name: string;
    description?: string;
    color?: string;
    icon?: string;
  },
): Promise<Workspace> {
  const response = await api.post(
    "/api/workspaces",
    data,
  );

  return response.data;
}

export async function updateWorkspace(
  id: number,
  data: {
    name?: string;
    description?: string;
    color?: string;
    icon?: string;
  },
): Promise<Workspace> {
  const response = await api.patch(
    `/api/workspaces/${id}`,
    data,
  );

  return response.data;
}

export async function deleteWorkspace(
  id: number,
) {
  await api.delete(
    `/api/workspaces/${id}`,
  );
}